File: 0046_Permutations/code4.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        
        n = len(nums)
        ans = []
        
        while not ans or ans[0] != nums:
            ans.append(nums.copy())
            
            i = n - 1
            while i - 1 >= 0 and nums[i - 1] > nums[i]:
                i -= 1
            i -= 1
            
            if i < 0:
                nums.reverse()
                continue
            
            j = n - 1
            while j - 1 >= 0 and nums[i] >= nums[j]:
                j -= 1

            nums[i], nums[j] = nums[j], nums[i]
            nums[i + 1:] = reversed(nums[i + 1:])

        return ans

File: 0046_Permutations/code4.cpp
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        int n = nums.size();
        
        vector<int> arr1 = nums;
        vector<vector<int>> ans;
        ans.push_back(nums);
        
        for(int k = 0; k < factorial(n) - 1; k++){
            vector<int> arr2 = arr1;
            
            int i = n - 2;
            while(i >= 0 && arr2[i] >= arr2[i + 1]) i--;            
            reverse(arr2.begin() + i + 1, arr2.end());
            
            if(i >= 0){
                int j = upper_bound(arr2.begin() + i + 1, arr2.end(), arr2[i]) - arr2.begin();
                swap(arr2[i], arr2[j]);
            }
            
            arr1 = arr2;
            ans.push_back(arr2);
        }
        
        return ans;
    }
    
    int factorial(int n){
        int ans = 1;
        for(int i = 1; i <= n; i++){
            ans *= i;
        }
        return ans;
    }
    
};